feat(56/STATUS-COMMUNITIES): initial draft (#567)
vacp2p/rfc: content/docs/rfcs/56/README.md

---
slug: 56
title: 56/STATUS-COMMUNITIES
name: Status Communities that run on Waku v2

> status: raw

Reviewer: This should be draft and correspond with actual code running Communities I think

Author: Addressed in [4d11e7a](https://github.com/vacp2p/rfc/pull/567/commits/4d11e7a584d384b7dee3267ee59964cf014a1eb5). It is *almost* 1:1 with the implementation, just a few features which are too large to mention here can be present in follow-up rfcs (backup/sync protocol, archival protocol, etc)

  // The community public key
  bytes community_id = 2;
}

message CommunityDescription {
  // The Lamport timestamp of the message
  uint64 clock = 1;
  // A mapping of members in the community to their roles
  map<string,CommunityMember> members = 2;
  // The permissions of the Community
  CommunityPermissions permissions = 3;
  // The metadata of the Community
  ChatIdentity identity = 5;
  // A mapping of chats to their details
  map<string,CommunityChat> chats = 6;
  // A list of banned members
  repeated string ban_list = 7;
  // A mapping of categories to their details
  map<string,CommunityCategory> categories = 8;
  // The admin settings of the Community
  CommunityAdminSettings admin_settings = 10;
  // If the community is encrypted
  bool encrypted = 13;
  // The list of tags
  repeated string tags = 14;
}
```

Note: The usage of the clock is described in the [Clock](#clock) section.

## Community Management

The flows for Community management are as described below.

### Community Creation Flow

1. The Community owner generates a public/private key pair.
2. The Community owner configures the Community metadata, according to the wire format "CommunityDescription".

> 3. The Community owner publishes the Community metadata on a content topic derived from the public key of the Community. 

Reviewer: More details on topic usage would be useful

Author: Added in [4d11e7a](https://github.com/vacp2p/rfc/pull/567/commits/4d11e7a584d384b7dee3267ee59964cf014a1eb5)

@felicio, can you confirm that the content topic usage is as mentioned in the commit? As per my investigation, this is how they're used.

Reviewer 2: Confirmed 👍, just

> keccak-256 hash of the result MUST be used as the content topic

the hash is encoded as hex and what's used as a topic is this https://rfc.vac.dev/spec/23/#bridging-waku-v1-and-waku-v2 construct.

Code for ref,

- https://github.com/status-im/status-go/blob/f6dc6f752a/wakuv2/common/topic.go#L66-L70
- https://github.com/status-im/status-go/blob/f6dc6f752ac87ff3f05e235aec18ad2bc36da791/protocol/transport/topic.go#L20-L23

> 32 bytes and used as the content topic

Same as above would apply.

Author: Great! thanks, I'll mention the hex-encoding

Author: Addressed in [bfd3e02](https://github.com/vacp2p/rfc/pull/567/commits/bfd3e021bddc2982f6964b5c85fcd1d97d9408e9)

---
slug: 56
title: 56/STATUS-COMMUNITIES
name: Status Communities that run on Waku v2
status: draft
category: Standards Track
tags: waku-application
editor: Aaryamann Challani <aaryamann@status.im>
contributors:
- Andrea Piana <andreap@status.im>
---

# Abstract

> This document describes the design of Status Communities for Waku v2, allowing for multiple users to communicate in a group chat. 

Reviewer 3: ```suggestion
This document describes the design of Status Communities over Waku v2, allowing for multiple users to communicate in a group chat. 
```

Reviewer 3: I'd say it is "over" because Waku v2 works at a level below Status communities.

This is a key feature for the Status messaging app. 

# Background and Motivation

Large group chats enable communities to communicate.
This would require channels, which are subject-based.
The messages in a channel are broadcasted to all the users in the channel.

A regular group chat between two or more peers reduces to a 1:1 chat between each peer and the other peers.

> One mechanism for 1:1 chats is described in [55/STATUS-1TO1-CHAT](/spec/55/).

Reviewer 3: Imo, the difference between a group chat and a community should be made clearer.
I assume, the current explanation would confuse somebody new to Status communities.

Maybe:
The purpose of Status communities, as specified in this document, is allowing for large group chats.
Communities can have further substructure, e.g. specific channels.

Smaller group chats, on the other hand, are out of scope for this document and can be  built over  [55/STATUS-1TO1-CHAT](/spec/55/). We refer to these smaller group chats simply as "group chats", to differentiate them from Communities. 
[continue with your explanation]

Author: Addressed in 72b855375bfcf94411bf3c48773b833679a6bc03

> The key-exchange mechanism MUST be X3DH, as described in [53/WAKU2-X3DH](/spec/53/).

Reviewer 3: For 1:1 chat?
If it MUST be X3DH, why does it leave implicit SHOULD or MAY above:

> One mechanism for 1:1 chats is described in

Maybe:
[...]  for group chats based on [55/STATUS-1TO1-CHAT](/spec/55/) the key-exchange mechanism MUST be X3DH...

Author: Addressed in 72b855375bfcf94411bf3c48773b833679a6bc03

> However, this method does not scale as the number of peers increases, for the following reasons -

Reviewer 3: ```suggestion
This method of building groups is not used for Communities, because it does not scale as the number of peers increases, for the following reasons
```

1. The number of messages sent over the network increases as the number of peers increases.
2. Handling the X3DH key exchange for each peer is computationally expensive.

> Having multicast channels reduces the overhead of a regular group chat.

Reviewer 3: ```suggestion
Having multicast channels reduces the overhead of a group chat based on 1:1 chat.
```

Additionally, if all the peers have a shared key, then the number of messages sent over the network is reduced to one per message.

# Design Requirements

Due to the nature of communities, the following requirements are necessary for the design of communities  -

> 1. The Community owner is trusted.

Reviewer 3: The concept of a community owner should be introduced first.

Reviewer 3: Is the owner a single enitiy, or can it be shared?

Author: Addressed in 64237af

2. The Community owner can add or remove peers from the Community.
This extends to banning and kicking peers.
3. The Community owner can add, edit and remove channels.
4. The peers in the Community can send/receive messages to the channels which they have access to.
5. Communities may be encrypted (private) or unencrypted (public).
6. A Community is uniquely identified by a public key.
7. The public key of the Community is shared out of band.
8. The metadata of the Community can be found by listening on a content topic derived from the public key of the Community.

> 9. Peers part of a Community run their own Waku nodes.

Reviewer 3: What does this mean?
Is it sufficient to run Waku v2 relay? Does every node have to run store (I assume not, but it should be explicit.)

Author: Addressed in 64237af

> Light peers may not be able to run their own Waku node.

Reviewer 3: I'd link to https://rfc.vac.dev/spec/10/ to explain what light peers are.
And I'd call them light node to be consistent.

Author: Addressed in 64237af

  bool accepted = 3;
  // The grant offered by the community
  bytes grant = 4;
  // The community public key
  bytes community_id = 5;
}

message CommunityRequestToLeave {
  // The Lamport timestamp of the request
  uint64 clock = 1;
  // The community public key
  bytes community_id = 2;
}

message CommunityDescription {
  // The Lamport timestamp of the message
  uint64 clock = 1;
  // A mapping of members in the community to their roles
  map<string,CommunityMember> members = 2;
  // The permissions of the Community
  CommunityPermissions permissions = 3;
  // The metadata of the Community
  ChatIdentity identity = 5;
  // A mapping of chats to their details
  map<string,CommunityChat> chats = 6;
  // A list of banned members
  repeated string ban_list = 7;
  // A mapping of categories to their details
  map<string,CommunityCategory> categories = 8;
  // The admin settings of the Community
  CommunityAdminSettings admin_settings = 10;
  // If the community is encrypted
  bool encrypted = 13;
  // The list of tags
  repeated string tags = 14;
}
```

Note: The usage of the clock is described in the [Clock](#clock) section.

## Content topic usage

### Advertising a Community

> The content topic that the community is advertised on MUST be derived from the public key of the community.

Reviewer 3: Link to content topic: https://rfc.vac.dev/spec/14/

Author: Addressed in 30c465bab5e82268d2a28fdcf08b4f37b667d896

The content topic MUST be the hex-encoded keccak-256 hash of the compressed (32 bytes) public key of the community.

```
contentTopic = hex(keccak256(compressedPublicKey))
```

### Community channels

The content topic that Community channels/chats use MUST be derived from the public key of the community and the chat id.

> The content topic MUST be the hex-encoded keccak-256 hash of the public key of the community concatenated with the chat id.

Reviewer 3: ```suggestion
The content topic that Community channels/chats use MUST be the hex-encoded keccak-256 hash of the public key of the community concatenated with the chat id.
```

Reviewer 3: I'd merge these two sentences. Imo, it is easier to grasp that way.
This more genereal statement:
> MUST be derived from the public key of the community and the chat id.
confuses, because it is restricted (and made superfluous) by the next statement.

```
contentTopic = hex(keccak256(compressedPublicKey + chatId))

> ```

Reviewer 3: Is `chatId` a `message ChatIdentity` ?

Author: After looking at the code, the `chatId` is a uuid generated when a chat needs to be created. Addressed in c0fc7cb

### Community event messages

Requests to leave, join, kick and ban, as well as key exchange messages, MUST be sent to the content topic derived from the public key of the community.
The content topic MUST be the hex-encoded keccak-256 hash of the public key of the community.

```
contentTopic = hex(keccak256(publicKey))

> ``` 

Reviewer 3: It might make sense to subdivide that space for specific control message types.
This would make content filtering more efficient.

Author: Addressed in fccad2d

## Community Management

The flows for Community management are as described below.

### Community Creation Flow

1. The Community owner generates a public/private key pair.
2. The Community owner configures the Community metadata, according to the wire format "CommunityDescription".
3. The Community owner publishes the Community metadata on a content topic derived from the public key of the Community. 

> he Community metadata SHOULD be encrypted with the public key of the Community. <!-- TODO: Verify this-->

Reviewer 3: ```suggestion
the Community metadata SHOULD be encrypted with the public key of the Community. <!-- TODO: Verify this-->
```

The Community metadata MAY be sent during fixed intervals, to ensure that the Community metadata is available to peers.
The Community metadata SHOULD be sent every time the Community metadata is updated.
4. The Community owner MAY advertise the Community out of band, by sharing the public key of the Community on other mediums of communication.

### Community Join Flow (peer requests to join a Community)

1. A peer generates a public/private key pair.

> 2. The peer requests to join a Community by sending a "CommunityRequestToJoin" message to the Community.

Reviewer 3: Note: We have to take this into consideration for DoS mitigation.
This message has to be relayed within the community, even though the sender is not part of the community.

At this point, the peer MAY send a "CommunityCancelRequestToJoin" message to cancel the request.
3. The Community owner MAY accept or reject the request.
4. If the request is accepted, the Community owner sends a "CommunityRequestToJoinResponse" message to the peer.
5. The Community owner then adds the member to the Community metadata, and publishes the updated Community metadata.

### Community Join Flow (peer is invited to join a Community)

> 1. The peer is invited to join a Community by the Community owner, by sending a "CommunityInvitation" message.

Reviewer 3: Limiting community access to this flow would solve the DoS issues I raised above.

Reviewer 4: I am guessing it's not sent out of band as it used a proto message. Is it sent using 1to1 chat protocol? If so, the peer need to already have a public/private key pair to receive the message via 1to1

Reviewer 5: It's sent via 1:1. The community owner should have the public key of the user being invited for the message to be broadcasted. Looking at how desktop works, to send an invite the community owner and the invitee must be contacts

Reviewer 3: We should add this to the spec @rymnc

Author: Addressed in 3690bab

2. A peer generates a public/private key pair.
3. The peer decrypts the "CommunityInvitation" message, and verifies the signature.
4. The peer requests to join a Community by sending a "CommunityRequestToJoin" message to the Community.
5. The Community owner MAY accept or reject the request.
6. If the request is accepted, the Community owner sends a "CommunityRequestToJoinResponse" message to the peer.
7. The Community owner then adds the member to the Community metadata, and publishes the updated Community metadata.

### Community Leave Flow

1. A peer requests to leave a Community by sending a "CommunityRequestToLeave" message to the Community.

> 2. The Community owner MAY accept or reject the request.

Reviewer 3: Why might a member not be allowed to leave :sweat_smile: ?

Author: As per the current implementation, the owner has control over this process in the community.

3. If the request is accepted, the Community owner removes the member from the Community metadata, and publishes the updated Community metadata.

### Community Ban Flow

1. The Community owner adds a member to the ban list, revokes their grants, and publishes the updated Community metadata.

## Waku protocols 

The following Waku protocols SHOULD be used to implement Status Communities -

1. [11/WAKU2-RELAY](/spec/11/) - To send and receive messages
2. [53/WAKU2-X3DH](/spec/53/) - To encrypt and decrypt messages
3. [53/WAKU2-X3DH-SESSIONS](/spec/54/) - To handle session keys
4. [14/WAKU2-MESSAGE](/spec/14/) - To wrap community messages in a Waku message
5. [13/WAKU2-STORE](/spec/13/) - To store and retrieve messages for offline devices

> 

Reviewer 3: Note: We should link the scaling BCP RFC here once it is done.

The following Waku protocols MAY be used to implement Status Communities -

1. [12/WAKU2-FILTER](/spec/12/) - Content filtering for resource restricted devices

## Backups

The peer MAY back up their local settings, by encrypting it with their public key, and sending it to a given content topic.
The peer MAY then rely on this backup to restore their local settings, in case of a data loss.
This feature relies on [13/WAKU2-STORE](/spec/13/) for storing and retrieving messages.

> 

Reviewer 3: We have to analyse this (or at least keep this in mind)  in regards to scaling.

## Clock

The clock used in the wire format refers to the Lamport timestamp of the message.
The Lamport timestamp is a logical clock that is used to determine the order of events in a distributed system.
This allows ordering of messages in an asynchronous network where messages may be received out of order.

> 

Reviewer 3: Note: This might duplicate features of Waku and add unnecessary complexty. Depends on how [Message ID](https://github.com/vacp2p/research/issues/175) evolves.

2. Handling the X3DH key exchange for each participant is computationally expensive.

Having multicast channels reduces the overhead of a group chat based on 1:1 chat.
Additionally, if all the participants of the group chat have a shared key, then the number of messages sent over the network is reduced to one per message.

# Terminology

- **Community**: A group of peers that can communicate with each other.
- **Member**: A peer that is part of a community.
- **Admin**: A member that has administrative privileges. Used interchangeably with "owner".
- **Channel**: A designated subtopic for a community. Used interchangeably with "chat".

# Design Requirements

Due to the nature of communities, the following requirements are necessary for the design of communities  -

1. The creator of the Community is the owner of the Community.
2. The Community owner is trusted.
3. The Community owner can add or remove members from the Community.
This extends to banning and kicking members.
4. The Community owner can add, edit and remove channels.
5. Community members can send/receive messages to the channels which they have access to.

> 6. Communities may be encrypted (private) or unencrypted (public).

Reviewer 4: My understanding is that public communities were dropped out. @john-44 can you confirm?

7. A Community is uniquely identified by a public key.
8. The public key of the Community is shared out of band.
9. The metadata of the Community can be found by listening on a content topic derived from the public key of the Community.
10. Community members run their own Waku nodes, with the configuration described in [#Waku-Protocols](#Waku-Protocols).

> Light nodes implementing [19/WAKU2-LIGHTPUSH](/spec/19/) may not be able to run their own Waku node with the configuration described.

Reviewer 4: I am a bit confused by this statement. Do we mean light nodes **solely** implementing light push?

Author: yes, Addressed in 53a7692

-->

```protobuf
syntax = "proto3";

message IdentityImage {
  // payload is a context based payload for the profile image data,
  // context is determined by the `source_type`
  bytes payload = 1;
  // source_type signals the image payload source
  SourceType source_type = 2;
  // image_type signals the image type and method of parsing the payload
  ImageType image_type = 3;
  // encryption_keys is a list of encrypted keys that can be used to decrypt an encrypted payload
  repeated bytes encryption_keys = 4;
  // encrypted signals the encryption state of the payload, default is false.
  bool encrypted = 5;
  // SourceType are the predefined types of image source allowed
  enum SourceType {
    UNKNOWN_SOURCE_TYPE = 0;

    // RAW_PAYLOAD image byte data
    RAW_PAYLOAD = 1;

    // ENS_AVATAR uses the ENS record's resolver get-text-data.avatar data
    // The `payload` field will be ignored if ENS_AVATAR is selected
    // The application will read and parse the ENS avatar data as image payload data, URLs will be ignored
    // The parent `ChatMessageIdentity` must have a valid `ens_name` set
    ENS_AVATAR = 2;
  }
}

> // SocialLinks represents social link assosiated with given chat identity (personal/community)

Reviewer 4: ```suggestion
// SocialLinks represents social link associated with given chat identity (personal/community)
```

}

message CommunityDescription {
  // The Lamport timestamp of the message
  uint64 clock = 1;
  // A mapping of members in the community to their roles
  map<string,CommunityMember> members = 2;
  // The permissions of the Community
  CommunityPermissions permissions = 3;
  // The metadata of the Community
  ChatIdentity identity = 5;
  // A mapping of chats to their details
  map<string,CommunityChat> chats = 6;
  // A list of banned members
  repeated string ban_list = 7;
  // A mapping of categories to their details
  map<string,CommunityCategory> categories = 8;
  // The admin settings of the Community
  CommunityAdminSettings admin_settings = 10;
  // If the community is encrypted
  bool encrypted = 13;
  // The list of tags
  repeated string tags = 14;
}
```

Note: The usage of the clock is described in the [Clock](#clock) section.

## Content topic usage

"Content topic" refers to the field in [14/WAKU2-MESSAGE](/spec/14/#message-attributes), further elaborated in [10/WAKU2](/spec/10/#overview-of-protocol-interaction).

### Advertising a Community

The content topic that the community is advertised on MUST be derived from the public key of the community.

> The content topic MUST be the hex-encoded keccak-256 hash of the compressed (32 bytes) public key of the community.

Reviewer 5: Maybe something like this?
```suggestion
The content topic MUST be the first four bytes of the keccak-256 hash of the compressed (33 bytes) public key of the community encoded into a hex string.
```

```

> contentTopic = hex(keccak256(compressedPublicKey))

Reviewer 5: ```suggestion
hash = hex(keccak256(encodeToHex(compressedPublicKey)))

topicLen = 4
if len(hash) < topicLen {
    topicLen = len(hash)
}

var topic [4]byte
for i = 0; i < topicLen; i++ {
    topic[i] = hash[i]
}

contentTopic = "/waku/1/0x" + topic + "/rfc26"
```
Using this as example: https://specs.status.im/spec/3#public-chats.
The important thing would be to mention somewhere that the compressed public key is handled as a hex string that starts with 0x (see https://github.com/status-im/status-go/blob/develop/protocol/communities/community.go#L1228) and that the contentTopic format follows what's described in https://rfc.vac.dev/spec/23/#differences-with-waku-v1

```

### Community channels/chats

The unique identifier for a community channel/chat is the chat id.
<!-- Don't enforce any constraints on the unique id generation -->
The content topic that Community channels/chats use MUST be the hex-encoded keccak-256 hash of the public key of the community concatenated with the chat id.

```

> contentTopic = hex(keccak256(compressedPublicKey + chatId))

Reviewer 5: See prev comment

Author: Addressed in 609fd24

```

### Community event messages

Requests to leave, join, kick and ban, as well as key exchange messages, MUST be sent to the content topic derived from the public key of the community.
The content topic MUST be the hex-encoded keccak-256 hash of the public key of the community.

```
contentTopic = hex(keccak256(publicKey))
``` 

## Community Management

The flows for Community management are as described below.

### Community Creation Flow

1. The Community owner generates a public/private key pair.
2. The Community owner configures the Community metadata, according to the wire format "CommunityDescription".
3. The Community owner publishes the Community metadata on a content topic derived from the public key of the Community. 

> the Community metadata SHOULD be encrypted with the public key of the Community. <!-- TODO: Verify this-->

Reviewer 4: Still a todo? Cc @felicio

Reviewer 2: The statement about metadata is true.

> The Community metadata MAY be sent during fixed intervals, to ensure that the Community metadata is available to members.

Reviewer 4: Would be good to define the current practice (what interval is used?) as this will help with the correctness of simulations.

Reviewer 5: I could not find instances of the community description being published periodically in status-go except when a change in the community description or control message happens.

Reviewer 6: https://github.com/status-im/status-go/blob/develop/protocol/messenger_communities.go#L191

Author: As per @felicio, https://discord.com/channels/864066763682218004/1065146990540902400/1067769073678680126
It's broadcasted every hour.

The Community metadata SHOULD be sent every time the Community metadata is updated.
4. The Community owner MAY advertise the Community out of band, by sharing the public key of the Community on other mediums of communication.

### Community Join Flow (peer requests to join a Community)

1. A peer generates a public/private key pair.
2. The peer requests to join a Community by sending a "CommunityRequestToJoin" message to the Community.
At this point, the peer MAY send a "CommunityCancelRequestToJoin" message to cancel the request.
3. The Community owner MAY accept or reject the request.
4. If the request is accepted, the Community owner sends a "CommunityRequestToJoinResponse" message to the peer.
5. The Community owner then adds the member to the Community metadata, and publishes the updated Community metadata.

### Community Join Flow (peer is invited to join a Community)

1. The peer is invited to join a Community by the Community owner, by sending a "CommunityInvitation" message.
2. A peer generates a public/private key pair.

> 3. The peer decrypts the "CommunityInvitation" message, and verifies the signature.

Reviewer 4: back to my comment above. How does the peer decrypt the message? I guess not with the key it just generated.

Author: Made it more clear in [3690bab](https://github.com/vacp2p/rfc/pull/567/commits/3690baba8f84cbe08393a7e1c1a514ab7f2e7683)

4. The peer requests to join a Community by sending a "CommunityRequestToJoin" message to the Community.
5. The Community owner MAY accept or reject the request.
6. If the request is accepted, the Community owner sends a "CommunityRequestToJoinResponse" message to the peer.
7. The Community owner then adds the member to the Community metadata, and publishes the updated Community metadata.

### Community Leave Flow

1. A member requests to leave a Community by sending a "CommunityRequestToLeave" message to the Community.
2. The Community owner MAY accept or reject the request.
3. If the request is accepted, the Community owner removes the member from the Community metadata, and publishes the updated Community metadata.

### Community Ban Flow

> 1. The Community owner adds a member to the ban list, revokes their grants, and publishes the updated Community metadata.

Reviewer 5: When an user is banned a re-key also happens if the community is encrypted. Should that be mentioned here?

Author: Addressed in b333c9c

## Waku protocols 

The following Waku protocols SHOULD be used to implement Status Communities -

1. [11/WAKU2-RELAY](/spec/11/) - To send and receive messages
2. [53/WAKU2-X3DH](/spec/53/) - To encrypt and decrypt messages
3. [53/WAKU2-X3DH-SESSIONS](/spec/54/) - To handle session keys
4. [14/WAKU2-MESSAGE](/spec/14/) - To wrap community messages in a Waku message
5. [13/WAKU2-STORE](/spec/13/) - To store and retrieve messages for offline devices

The following Waku protocols MAY be used to implement Status Communities -

> 1. [12/WAKU2-FILTER](/spec/12/) - Content filtering for resource restricted devices

Reviewer 4: Why is light push not mentioned here?

Author: Addressed in 53a7692
